---
title: "1.03_Gather-and-Tidy-Data"
author: ""
date: "`r format(Sys.time(), '%d %B %Y')`"
output:
  rmarkdown::html_document:
    toc: TRUE
---

### Install/Load dependencies
```{r, echo=TRUE, message=FALSE, Loading_Dependencies_1}
# Load dependencies:
library(BiocManager)
library(Seurat)
library(SeuratData)
library(patchwork)
```

## Read-in input data:

For convenience, Seurat distributes this dataset through the SeuratData package.
```{r} 
# load dataset
InstallData("ifnb")
LoadData("ifnb")
```

## Metadata from Seurat
Seurat automatically creates some metadata for each of the cells:
```{r}
head(ifnb@meta.data)
```
The columns include:

1. orig.ident: this often contains the sample identity if known, but will default to project as we had assigned it
2. nCount_RNA: number of UMIs per cell
3. nFeature_RNA: number of genes detected per cell

We need to calculate some additional metrics for plotting:

1. number of genes detected per UMI: this metric with give us an idea of the complexity of our dataset (more genes detected per UMI, more complex our data)
2. mitochondrial ratio: this metric will give us a percentage of cell reads originating from the mitochondrial genes - this was already done by Lyko in this case, but I'll calculate again.

The number of genes per UMI for each cell is quite easy to calculate, and we will log10 transform the result for better comparison between samples.

### Calculate number of genes per UMI for each cell to metadata
```{r calculate_nUMI}
ifnb$log10GenesPerUMI <- log10(ifnb$nFeature_RNA) / log10(ifnb$nCount_RNA)
```

Seurat has a convenient function that allows us to calculate the proportion of transcripts mapping to mitochondrial genes. The PercentageFeatureSet() will take a pattern and search the gene identifiers. For each column (cell) it will take the sum of the counts slot for features belonging to the set, divide by the column sum for all features and multiply by 100. Since we want the ratio value for plotting, we will reverse that step by then dividing by 100.

### Compute percent mito ratio:
Seurat has a convenient function that allows us to calculate the proportion of transcripts mapping to mitochondrial genes. The PercentageFeatureSet() function takes in a pattern argument and searches through all gene identifiers in the dataset for that pattern. Since we are looking for mitochondrial genes, we are searching any gene identifiers that begin with the pattern "MT-". For each cell, the function takes the sum of counts across all genes (features) belonging to the "Mt-" set, and then divides by the count sum for all genes (features). This value is multiplied by 100 to obtain a percentage value.

For our analysis, rather than using a percentage value we would prefer to work with the ratio value. As such, we will reverse that last step performed by the function by taking the output value and dividing by 100.

```{r} 
# The [[ operator can add columns to object metadata. This is a great place to stash QC stats
ifnb[["percent.mt"]] <- PercentageFeatureSet(ifnb, pattern = "^MT-")
ifnb[["mitoRatio"]] <- ifnb@meta.data$percent.mt / 100
```


### Session information
```{r session_info_103}
sessionInfo()
```


This document was processed on: `r Sys.Date()`.
